ProcessLib/ThermoMechanicalPhaseField/ThermoMechanicalPhaseFieldFEM-impl.h

```diff
@@ -0,0 +1,644 @@
+
+    auto local_rhs =
+        MathLib::createZeroedVector<typename ShapeMatricesType::
+                                        template VectorType<displacement_size>>(
+            local_b_data, displacement_size);
+
+    double const& dt = _process_data.dt;
+
+    SpatialPosition x_position;
+    x_position.setElementID(_element.getID());
+
+    int const n_integration_points = _integration_method.getNumberOfPoints();
+    for (int ip = 0; ip < n_integration_points; ip++)
+    {
+        x_position.setIntegrationPoint(ip);
+        auto const& w = _ip_data[ip].integration_weight;
+        auto const& N = _ip_data[ip].N;
+        auto const& dNdx = _ip_data[ip].dNdx;
+
+        auto const x_coord =
+            interpolateXCoordinate<ShapeFunction, ShapeMatricesType>(_element,
+                                                                     N);
+        auto const& B =
+            LinearBMatrix::computeBMatrix<DisplacementDim,
+                                          ShapeFunction::NPOINTS,
+                                          typename BMatricesType::BMatrixType>(
+                dNdx, N, x_coord, _is_axially_symmetric);
+
+        auto& eps = _ip_data[ip].eps;
+
+        auto const& C_tensile = _ip_data[ip].C_tensile;
+        auto const& C_compressive = _ip_data[ip].C_compressive;
+
+        auto const& sigma = _ip_data[ip].sigma;
+
+        auto const k = _process_data.residual_stiffness(t, x_position)[0];
+        auto rho_sr = _process_data.solid_density(t, x_position)[0];
+        auto const alpha = _process_data.linear_thermal_expansion_coefficient(
+            t, x_position)[0];
+        double const T0 = _process_data.reference_temperature;
+        auto const& b = _process_data.specific_body_force;
+
+        double const T_ip = N.dot(T);
+        double const delta_T = T_ip - T0;
+        // calculate real density
+        double const rho_s = rho_sr * (1 - 3 * alpha * delta_T);
+
+        double const d_ip = N.dot(d);
+        double const degradation = d_ip * d_ip * (1 - k) + k;
+
+        auto const C_eff = degradation * C_tensile + C_compressive;
+        eps.noalias() = B * u;
+        _ip_data[ip].updateConstitutiveRelation(t, x_position, dt, u, alpha,
+                                                delta_T, degradation);
+
+        typename ShapeMatricesType::template MatrixType<DisplacementDim,
+                                                        displacement_size>
+            N_u = ShapeMatricesType::template MatrixType<
+                DisplacementDim, displacement_size>::Zero(DisplacementDim,
+                                                          displacement_size);
+
+        for (int i = 0; i < DisplacementDim; ++i)
+            N_u.template block<1, displacement_size / DisplacementDim>(
+                   i, i * displacement_size / DisplacementDim)
+                .noalias() = N;
+
+        using Invariants = MathLib::KelvinVector::Invariants<
```

Comment:
Unused typedef.

Reply:
removed all unused functions and definitions.